### DISCUSSION TITLE
BDCS - (One of) the first dynamic kernel libraries.

### DISCUSSION THREAD

**Discussion by ghost** [CosmosOS/Cosmos#1870] [Show and tell]
**[BDCS](https://github.com/BeriiDevelopment/bdcs) or "Berii Dynamic Command System" is a NuGet package mainly designed for CosmosOS that is used to create dynamic kernels and command systems.**

***

**About**

This project was based off of [BeriiKernel](https://github.com/BeriiDevelopment/bknl) but has then branched out into a separate project.
It's fully customizable, with custom functions and actions for all classes and objects including CommandHandlers, CommandListeners, and CommandActions.

Although the full BDCS library doesn't work in Cosmos yet due to unsupported use of Func<string, string[], bool> objects, this is a working concept. I've proven this concept working in [this](https://github.com/BeriiDevelopment/bknl) GitHub repo. It makes use of an older, working version of BDCS using Actions intead of Funcs. It's not as customizable though.

If you want the new BDCS library to work, help solve [this](https://github.com/CosmosOS/Cosmos/issues/1869) issue!

***

**How to use**

Currently, you can only use this library in regular .NET standard 2.0 projects. **If the issue above is solved though, this will already work with the devkit released if no other problems occur when compiling my test kernel**.

Here's a custom command listener!
```cs
CommandInfo cmd  = new CommandListener().Listen(() => {
  /// This is a command listener with a custom function. This reads a line from the console and splits it into a CommandInfo object.
  string line = Console.ReadLine();
  return new CommandInfo(line.Split(' ')[0], line.Split(' '));
});
```

Want to make a custom command handler? Look no further!
```cs
new CommandHandler().HandleCommand(cmd.firstCommand, cmd.fullCommand, (firstCommand, fullCommand) => {
  /// This action dynamically handles a command.
  bool validCommand = false;

  foreach (Command command in CommandManager.commands)
    if (firstCommand == command.name)
      validCommand = command.action.Invoke(firstCommand, fullCommand);

    if (!validCommand)
      Console.WriteLine("Invalid command.");
});
```

Want to simplify making a command?
```cs
new CommandManager().RegisterCommand(new Command("help", "shows help menu", (firstCommand, fullCommand) => {
  foreach (Command command in CommandManager.commands)
    Console.WriteLine($"{command.name} - {command.description}");
  return true;
}));
```

I tried to keep this as easy as possible to ease command creation. Using lambda expressions, you can make everything extremely simple.

**Comment by dm-vev** [comment]
Great project! Thanks
